relays/ethereum/src/ethereum_exchange.rs

```diff
@@ -0,0 +1,212 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::ethereum_client::{EthereumConnectionParams, EthereumRpcClient};
+use crate::ethereum_types::{
+	EthereumHeaderId, Transaction as EthereumTransaction, TransactionHash as EthereumTransactionHash, H256,
+};
+use crate::exchange::{relay_single_transaction_proof, SourceClient, TargetClient, TransactionProofPipeline};
+use crate::rpc::{EthereumRpc, SubstrateRpc};
+use crate::rpc_errors::{EthereumNodeError, RpcError};
+use crate::substrate_client::{
+	SubmitEthereumExchangeTransactionProof, SubstrateConnectionParams, SubstrateRpcClient, SubstrateSigningParams,
+};
+use crate::sync_types::HeaderId;
+
+use async_trait::async_trait;
+use bridge_node_runtime::exchange::EthereumTransactionInclusionProof;
+use std::time::Duration;
+
+/// Interval at which we ask Ethereum node for updates.
+const ETHEREUM_TICK_INTERVAL: Duration = Duration::from_secs(10);
+/// Interval at which we ask Substrate node for updates.
+const SUBSTRATE_TICK_INTERVAL: Duration = Duration::from_secs(5);
+
+/// PoA exchange transaction relay params.
+#[derive(Debug, Default)]
+pub struct EthereumExchangeParams {
+	/// Ethereum connection params.
+	pub eth: EthereumConnectionParams,
+	/// Hash of the Ethereum transaction to relay.
+	pub eth_tx_hash: EthereumTransactionHash,
+	/// Substrate connection params.
+	pub sub: SubstrateConnectionParams,
+	/// Substrate connection params.
+	pub sub_sign: SubstrateSigningParams,
+}
+
+/// Ethereum to Substrate exchange pipeline.
+struct EthereumToSubstrateExchange;
+
+impl TransactionProofPipeline for EthereumToSubstrateExchange {
+	const SOURCE_NAME: &'static str = "Ethereum";
+	const TARGET_NAME: &'static str = "Substrate";
+
+	type BlockHash = H256;
+	type BlockNumber = u64;
+	type TransactionHash = EthereumTransactionHash;
+	type Transaction = EthereumTransaction;
+	type TransactionProof = EthereumTransactionInclusionProof;
+}
+
+/// Ethereum node as transactions proof source.
+struct EthereumTransactionsSource {
+	client: EthereumRpcClient,
+}
+
+#[async_trait]
+impl SourceClient<EthereumToSubstrateExchange> for EthereumTransactionsSource {
+	type Error = RpcError;
+
+	async fn tick(&self) {
+		async_std::task::sleep(ETHEREUM_TICK_INTERVAL).await;
+	}
+
+	async fn transaction(
+		&self,
+		hash: &EthereumTransactionHash,
+	) -> Result<Option<(EthereumHeaderId, EthereumTransaction)>, Self::Error> {
+		let eth_tx = match self.client.transaction_by_hash(*hash).await? {
+			Some(eth_tx) => eth_tx,
+			None => return Ok(None),
+		};
+
+		// we need transaction to be mined => check if it is included in the block
+		let eth_header_id = match (eth_tx.block_number, eth_tx.block_hash) {
+			(Some(block_number), Some(block_hash)) => HeaderId(block_number.as_u64(), block_hash),
+			_ => return Ok(None),
```

Comment:
How can we get to this point? If a transaction hasn't been included in a block wouldn't the call to `transaction_by_hash()` return `None`?

Reply:
`None` is returned when transaction is completely unknown to the node. But if it is in the tx pool, it would rather return tx with these fields set to `None`.